core/state_transition.go

```diff
@@ -242,6 +254,64 @@ func NewStateTransition(evm *vm.EVM, msg *Message, gp *GasPool) *StateTransition
 	}
 }
 
+func (st *StateTransition) checkGasFormula() error {
+	if st.boughtGas.Cmp(
+		new(uint256.Int).Add(
+			st.refundedGas, new(uint256.Int).Add(
+				st.tipFee, new(uint256.Int).Add(
+					st.baseFee, st.l1Fee)))) != 0 {
+		return fmt.Errorf("gas formula doesn't hold: boughtGas(%v) != refundedGas(%v) + tipFee(%v) + baseFee(%v) + l1Fee(%v)", st.boughtGas, st.refundedGas, st.tipFee, st.baseFee, st.l1Fee)
+	}
+	return nil
+}
+
+// distributeGas distributes the gas according to the priority:
+//  1. if refund, first native, then SGT
+//  2. otherwise, first SGT, then native
+//
+// note: the returned values are all non-nil.
+func (st *StateTransition) distributeGas(amount *uint256.Int, refund bool) (native, sgt *uint256.Int) {
```

Comment:
you can simplify the call as 
```
func (st) distributeGas(amount *uint256.Int, bal1 *uint256.Int, bal2 *uint256.Int) (...) { // Assume no of them are nil
   if amount.Cmp(bal1Bought) >= 0 {
      bal2Reduced = amount.Sub(amount.Clone(), bal1)
      bal2.Sub(bal2Bought, bal2)
      bal1Reduced := bal1.Clone()
      bal1.Clear()
      return bal1Reduced, bal2Reduced
   } else {
      bal1.Sub(bal1, amount)
      return amount.Clone(), 0
   }
}
```

And then the caller can call `distributeGas(...)` with different refund order.

Reply:
That's a really good idea! Done in c84608573c6d9f62ac331f1c9c69a5cfb2a65b3c.